### PR TITLE
Internal Refactoring for 3.4.3

### PR DESCRIPTION
**Pull request by StandingPadAnimations**
# List of Deprecated Features Planned for MCprep 3.4.3
## `conf.dev` -> `conf.env.dev_build`
```py
from .conf import env
env.dev_build
```

## `conf.v` -> `conf.env.verbose`
```py
from .conf import env
env.verbose
```

## `conf.vv` -> `conf.env.very_verbose`
```py
from .conf import env
env.very_verbose
```

## `conf.data` -> `conf.env.data`
```py
from .conf import env
env.data
```

## `conf.json_data` -> `conf.env.json_data`
```py
from .conf import env
env.json_data
```

## `conf.json_path` -> `conf.env.json_path`
```py
from .conf import env
env.json_path
```

## `conf.use_icons` -> `conf.env.use_icons`
```py
from .conf import env
env.use_icons
```

## `conf.preview_collections` -> `conf.env.preview_collections`
```py
from .conf import env
env.preview_collections
```

## `conf.icons_init()` -> `conf.env.icons_init()`
```py
from .conf import env
env.icons_init()
```

## `conf.loaded_all_spawners` -> `conf.env.loaded_all_spawners`
```py
from .conf import env
env.loaded_all_spawners
````

## `conf.skin_list` -> `conf.env.skin_list`
```py
from .conf import env
env.skin_list 
```

## `conf.rig_categories` -> `conf.env.rig_categories`
```py
from .conf import env
env.rig_categories 
```

## `conf.entity_list` -> `conf.env.entity_list`
```py
from .conf import env
env.entity_list
```

## `conf.material_sync_cache` -> `conf.env.material_sync_cache`
```py
from .conf import env
env.material_sync_cache
```

## `conf.register()` ->`conf.MCprepEnv.__init__(self)`
```py
# conf.py
env = MCprepEnv()
```

## `conf.log(statement, vv_only=False)` -> `conf.env.log(self, statement, vv_only=False)`
```py
from .conf import env
env.log("Debug statement!")
```

### CODE REVIEW

**Comment by TheDuckCow** [comment]
Realize this is still in review, but calling out now that I'd prefer the name lower-case ENV instead of uppercase; uppercase is generally referred to treat something as a constant, but env is something that will be mutable (verbose can be turned off and on for instance), so will ask you to turn it into lowercase when you can.

Holding off on further review in the meantime - thanks for getting a head start here!

**Comment by TheDuckCow** [comment]
Yup that makes sense to have those vars initialized in `__init__`; we can do a number of interesting things indeed, though we want to make sure that it stays lightweight.

As for the name of the variable instance, I'd like a convention for naming that is short and easy to use in the other classes. `mcprep_env` feels a bit long, I feel like we could use `env` since it's not used anywhere else. We don't have any env files at the moment, and even if we did, it would be read and parse by the MCprepEnv class anyways, so it feels accurate enough that you would call the instance overall as env anyhow.

**Comment by StandingPadAnimations** [comment]
Sorry for the lack of updates, hopefully I'll be able to get some changes done by late December/early January. I'll be busy for the next couple of weeks

**Comment by StandingPadAnimations** [comment]
Alright, @TheDuckCow, I've renamed `ENV` to `env`. Hopefully by the end of this week I can get some more progress done on this

**Comment by StandingPadAnimations** [comment]
As mentioned [here](https://standingpadanimations.github.io/posts/goodbye-for-now/), I won't be working on MCprep anymore, so this PR will just be left in this state.

I believe this will eventually get done in the future, either by the open source community or myself (if I decide to come back) but for now I'll just set this as a draft

**Comment by StandingPadAnimations** [comment]
Ok, so I don't know if I'll reopen this, but I'm curious about whether we'll use type annotations (which are supported in Blender 2.8 and above), since it would give the following benefits:
* Better completions in IDES
* More info for static analyzers, which MCprep would massively benefit from
* Tells developers what arguments a function can take and return
* etc
Alternatively we could add type annotations at a later date (not preferable, but better late then never) or figure out a way to use it in older versions of Blender (perhaps with the futures module?)

**Comment by StandingPadAnimations** [comment]
I think now I'll be working on this. I'll look into porting the use of OS to Pathlib (for file operations) since it's more readable and concise

**Comment by StandingPadAnimations** [comment]
Though now I got to sign all of these commits for verification purposes. Not too annoying though since it's not merged

**Comment by StandingPadAnimations** [comment]
Alright with #399 (and this PR is for MCprep 3.5), I think it's now safe to deprecate the use of `os` for filepaths and file operations, as well as add type annotations to the env class

**Comment by StandingPadAnimations** [comment]
Alright, I'll merge dev with this locally (for updates) ajd get started. I'll be doing the following:
- Add a deprecation warning function
- Deprecate 2.7x compatability code
- Remove 2.7x compatablity code where possible
- Add type hints to functions and variables

@TheDuckCow just want to know your thoughts

**Comment by TheDuckCow** [comment]
Just to be clear, you mean merge `dev` *into* this branch just to sync it up right?

I'm ok with this. We can use internal-rewrites as an actual "feature branch". I don't want to merge breaking compatibility into dev yet, as we may have more patch releases between now and 3.5 (we will for sure have at least a 3.5.2, based on your report about mtl conversion errors). But it would be good to get a head start on some of these updates.

I'd start with removing 2.7x compatibility code, anything to do with Blender Internal.

I'm wary about adding type hints everywhere *just yet*. It will just create more merge conflicts in the future if we annotate every function, whereas if we do this a little closer to the 3.5 release, we'd deal with the state of the code at that time. But this could also be mitigated by diligently re-merging dev into this `internal-rewrites` feature branch.

**Comment by StandingPadAnimations** [comment]
> Just to be clear, you mean merge dev into this branch just to sync it up right?

Yep

> I'd start with removing 2.7x compatibility code, anything to do with Blender Internal.

Sure thing

**Comment by StandingPadAnimations** [comment]
Btw @TheDuckCow, how is fast compile performed? Since both MacOS and Linux are UNIX, I'm curious if I can get it to work on Linux (unless it already does)

**Comment by StandingPadAnimations** [comment]
Crap, did a rebase wrong

**Comment by StandingPadAnimations** [comment]
there we go, fixed it

**Comment by StandingPadAnimations** [comment]
Alright, removed the Blender Internal code (at least for prep materials). Even [UPBGE](https://upbge.org/) (a fork of the Blender Game Engine with improvements based on newer versions of Blender) doesn't use Blender Internal and instead uses EEVEE, so support won't be missed

File: MCprep_addon/conf.py
Comment (TheDuckCow) [inline MCprep_addon/conf.py:20]: FYI: I just sanity confirmed this imports fine in blender 2.79 ok

File: MCprep_addon/util.py
Comment (TheDuckCow) [inline MCprep_addon/util.py:563]: Suggested docstring. Also, can we gate it so that it exits early (with no prints or anything) if it's in prod mode? This way we aren't polluting console logs for the next smaller couple releases of MCprep.

```suggestion
def deprecation_warning():
	"""Creates printouts in functions which will be removed on the v3.5 MCprep release"""
	print("DEPRECATION WARNING: This function will be deprecated in MCprep v3.5")
```

File: MCprep_addon/spawner/spawn_util.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/spawn_util.py:819]: In this file here and below: If we want to merge this to dev now, then we should revert the use of annotations. Use of annotations and typing should go into the (not yet existing) `MCprep_v350` feature branch instead. Otherwise it would be an immediate drop of support for Blender 2.7x.

File: MCprep_addon/spawner/meshswap.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/meshswap.py:111]: I wonder if we should just create a .debug function instead, and nix the use of a vv_only flag.

File: MCprep_addon/spawner/item.py
Comment (TheDuckCow) [inline MCprep_addon/spawner/item.py:300]: Same as commented in other files: Let's hold off on annotation changes for now.

File: MCprep_addon/conf.py
Comment (TheDuckCow) [inline MCprep_addon/conf.py:308]: Per comment somewhere lower, maybe we should just have a second `def debug` function, instead of a log with a vv_only flag.

**Review by TheDuckCow — CHANGES_REQUESTED** [review]
Initial round of commnets, didn't get to everything. Main thing is to hold off on annotations, and I am noticing tests are failing for me locally. I can look into more later if you'd like me to (have you be able to run the test suite before?)

File: MCprep_addon/spawner/spawn_util.py
Comment (StandingPadAnimations) [reply MCprep_addon/spawner/spawn_util.py:819]: I'm planning on holding off until 3.5 to merge this

**Comment by StandingPadAnimations** [comment]
Yeah I need to update the comments, I plan to merge this for MCprep 3.5

**Comment by StandingPadAnimations** [comment]
Yeah one problem with the test suite:
`./run_tests.sh: ./mcprep_data_refresh.py: /opt/homebrew/bin/python3: bad interpreter: No such file or directory`

On Linux we don't use Homebrew, we just install from our package manager. For me (on Arch Linux), that would be `/usr/bin/python3`

**Comment by StandingPadAnimations** [comment]
Yeah this will be a problem (I don't have Minecraft installed on this laptop because it's an absolute nightmare on Linux to install)
image

**Comment by StandingPadAnimations** [comment]
Alright, so I think I'll just revert the 2.7x changes (or move them to a different branch for later) and just keep this for deprecations

**Comment by StandingPadAnimations** [comment]
Alright, since it's been a bit confusing the past couple of days, I've went ahead and changed the plans slightly for this branch. 

For starters, I'll be removing the commits made yesterday. Here's how that will look:
## 3.4.3 (assuming this is the last in the 3.4 series, correct me if I'm wrong) 
Replace conf globals with MCprepEnv, basically what was planned originally. In addition, I'll also be integrating some new build stuff (once I simplify it) 

### 3.5 
Remove previously deprecated code, deprecate compatability code (well at least some of it, other parts are actually pretty useful readability wise), begin moving to annotations (for property groups, commonly used functions, and MCprepEnv)

### 3.5.1 
Remove deprecated compatability code, start adding annotations for everything else, just continue as normal

@TheDuckCow just nees your thoughts on this new plan

**Comment by TheDuckCow** [comment]
I'm in line with this plan - but do be safe, until we are _very close_ to v3.5, we should maintain a 3.5 branch so that dev could turn into a 3.4.4 if necessary.

Thanks for helping organize this @StandingPadAnimations

**Comment by StandingPadAnimations** [comment]
> we should maintain a 3.5 branch so that dev could turn into a 3.4.4 if necessary.

Sure thing, I'll make one based off of this branch and cherry-pick yesterday's commits there

**Comment by StandingPadAnimations** [comment]
Alright, went ahead and made a new branch called `mcprep-3_5-recore` (recore being short for rewritten core, as many parts are being rewritten)

**Comment by StandingPadAnimations** [comment]
So should we move these deprecations to MCprep 3.5?

**Comment by TheDuckCow** [comment]
I'm fine to proceed with merging these changes to a dedicated MCprep 3.5 branch (not Dev, but newly branched from current dev)Let me set this up later when I have time (give me a day or so). Idea is it will be a branch we treat similar to dev. We merge approved PR changes that are 3.5x only into that, and then only when we are totally ready to move ahead with 3.5x code (ie we know we won't need any further patch releases), then we merge this mcprep35 branch into dev. Until then, I'll be diligent and merge dev changes back into this mcprep35 branch so it doesn't get too out of sync.On Apr 6, 2023 9:45 AM, StandingPad ***@***.***> wrote:
So should we move these deprecations to MCprep 3.5?

**Comment by TheDuckCow** [comment]
I've updated the base since this branch as-is doesn't work in 2.7x (first error is `print(f"\"{header}\"")`), and this is ok. I'm hopeful we can make 3.5.0 the next release anyways.

Some changes still needed though, errors on startup with I think some remaining `ENV` variables where the class instance is now named `env` (which I think is good). e.g. 


```
  File "/Users/patrickcrawford/Library/Application Support/Blender/3.5/scripts/addons/MCprep_addon/mcprep_ui.py", line 934, in draw
    elif env.skin_list and len(ENV.skin_list) <= sind:
NameError: name 'ENV' is not defined. Did you mean: 'env'?
```

and

```
Traceback (most recent call last):
  File "/Applications/Blender 3.5/blender.app/Contents/Resources/3.5/scripts/modules/addon_utils.py", line 369, in enable
    mod.register()
  File "/Users/patrickcrawford/Library/Application Support/Blender/3.5/scripts/addons/MCprep_addon/__init__.py", line 64, in register
    load_modules.register(bl_info)
  File "/Users/patrickcrawford/Library/Application Support/Blender/3.5/scripts/addons/MCprep_addon/load_modules.py", line 181, in register
    addon_updater_ops.register(bl_info)
  File "/Users/patrickcrawford/Library/Application Support/Blender/3.5/scripts/addons/MCprep_addon/addon_updater_ops.py", line 1360, in register
    from .conf import ENV
ImportError: cannot import name 'ENV' from 'MCprep_addon.conf' (/Users/patrickcrawford/Library/Application Support/Blender/3.5/scripts/addons/MCprep_addon/conf.py)
```
And some others.

Once we resolve these, I think it would be good to merge into the newly-assigned feature branch, so that other changes can be reviewed on their own.

**Comment by StandingPadAnimations** [comment]
>    I've updated the base since this branch as-is doesn't work in 2.7x
>    (first error is print(f"\"{header}\"")), and this is ok. I'm hopeful we
>    can make 3.5.0 the next release anyways.
I think we can ignore that if we merge this for MCprep 3.5

>    Some changes still needed though, errors on startup with I think some
>    remaining ENV variables where the class instance is now named env
>    (which I think is good). e.g.
Whoops, looks like I missed some places, good catch

>    Once we resolve these, I think it would be good to merge into the
>    newly-assigned feature branch, so that other changes can be reviewed on
>    their own.
The MCprep 3.5 Recore branch or a new branch? If the latter, it might be
easier to just keep them here.

**Comment by StandingPadAnimations** [comment]
Now that there's a milestone branch for 3.5, should we merge this branch?

**Comment by TheDuckCow** [comment]
> Now that there's a milestone branch for 3.5, should we merge this branch?

Let's fix the couple things mentioned where you missed first, and resolve the conflicts. I'll take one more quick look then, and then yes we can merge into the milestone branch.

**Comment by StandingPadAnimations** [comment]
Alright, fixed the remaining uses of `ENV` and also removed the use of `conf.load()` in `load_modules.py` (I love the new deprecation warnings I've added). I've also updated the deprecation warning to say `3.5.1` (I'm 50% we'll just push back the removal, but at the very least we should start throwing around deprecation warnings)

@TheDuckCow just need some final checks on this to make sure everything is working as expected

**Comment by StandingPadAnimations** [comment]
Alright I've now pushed the dev build feature to `compile.sh`. Just know that `-fast` has also now become `-f`

**Comment by StandingPadAnimations** [comment]
Ok I just realized that we also have #401, which makes this branch a bit weird (#401 is also based off of this). Should we merge this completely with `recore` or should we merge this with the milestone branch and then merge `recore` later on?

I'm leaning more towards the latter so that the changes are more incremental, but I want to know everyone's thoughts

**Comment by TheDuckCow** [comment]
> Ok I just realized that we also have #401, which makes this branch a bit weird (#401 is also based off of this). Should we merge this completely with `recore` or should we merge this with the milestone branch and then merge `recore` later on?
> 
> I'm leaning more towards the latter so that the changes are more incremental, but I want to know everyone's thoughts

It's not a problem, both can be merged into milestone-3-5-0, in whichever order (both will likely have a couple merge conflicts, so dealers choice which you want to deal with in which order). So I guess the latter of your two options is what I agree with. The sooner we merge this one though, the fewer problems we'll have. If we need to make future fixes, they can be branched off of the milestone-3-5-0 and PR-merged back into the same milesotne-3-5-0

**Comment by StandingPadAnimations** [comment]
Closing this as this is no longer needed with #401

**Pull request closed without merging**
